### DISCUSSION TITLE
Install pkgs w/ dependencies managed by asdf?

### DISCUSSION THREAD

**Discussion by TemporaryEfx** [Homebrew/discussions#1621] [Everyday usage] [answered]
**Question:** When Homebrew checks for installed dependencies, it is possible to direct Homebrew to broaden its search to include languages managed by asdf? 

**Scenario:** I installed [asdf](https://github.com/asdf-vm/asdf) via Homebrew. I currently use asdf to install and manage runtimes of languages including Go, Python, Ruby, Sqlite, etc. Now, whenever I `brew install ...` a new formulae that depends on one of my asdf-managed languages, Homebrew flags the language as a missing dependency. This occurs despite the fact that there is a fully operational and globally available version of that language already installed on my system.  I can work around this problem by manually brewing each dependency and skipping the existing language. This results in a working installation, but it is a time consuming process. Even after initial installation of the new formulae, `brew doctor` and `brew missing` continue to (incorrectly) report the language as an absent dependency. 

**Desired Solution:** It'd be great if I could find a way to either: (a) Make Homebrew cognizant of asdf's [global version](https://asdf-vm.com/#/core-manage-versions?id=set-current-version) of an installed language whenever Homebrew checks for the existence of a formulae's dependencies; or (b) Get more accurate output from brew doctor and brew missing, such as: "Some installed formulae ~~are missing dependencies~~ _have dependencies that are not in your Brew List_." 

Thanks to everyone here for all of the helpful advice in this community... Any assistance you all can provide will be greatly appreciated! 

P.S. Here's how the folks on the asdf side of the fence [responded](https://github.com/asdf-vm/asdf/discussions/869) to a similar question.
P.P.S. A _very_ similar question was [posted](https://stackoverflow.com/questions/61530830) on Stack Overflow more than a year ago, but it remains open and unanswered.

**Comment by nandahkrishna** [comment, marked as answer]
This is similar to https://github.com/Homebrew/brew/issues/11452 and https://github.com/Homebrew/brew/issues/11312, this feature request has been made many times before but the responses there (and in other similar issues) could help you understand why we don't do this.

**Reply by TemporaryEfx** [reply]
Thank you for such a quick reply! It looks like what I am asking about may be somewhat different from (and more achievable than?) the prior questions you referenced. Nonetheless, I now understand that what I'm about to say is actually a feature request, not a usage question, so I'd be happy to close this discussion and move there if you feel that is appropriate. 

Could `Brew Install`, `Brew Doctor`, and `Brew Missing` include an option (perhaps "--locate-dependencies") such that:
- IF the option is selected AND IF the Brew List does _**not**_ contain a required dependency
   - THEN Homebrew executes a simple `which [dependency]` (or `type [dependency]`) shell command 
   - IF the shell command returns any valid result 
      - THEN Homebrew recognizes the presence of the dependency and does not attempt to install it or otherwise flag it as missing
     - ELSE the shell will return a null result (or "Could not find [dependency]" when the `type` command is used) in which case Homebrew can treat the dependency as missing just as it does now.
- The description text for this proposed option could start with the same warning currently found on the --ignore-dependencies option: "An unsupported Homebrew development flag...". 

Irrespective of whether or not the above proposal is feasible, it would be useful to revise the `Brew Doctor` (and `Brew Missing`?) output string to more accurately describe the information that Homebrew has––and does not have!––about the applications that exist on the system. Specifically, in the scenario described above, `Brew Doctor` currently provides the following inaccurate / misleading information and advice: 

> Warning: Some installed formulae are missing dependencies.
> You should `brew install` the missing dependencies:
>   brew install [dependency]

This could be easily corrected by amending the output string as follows: 

> Warning: Some installed formulae have dependencies that are not found in your Brew List. 
> If these dependencies exist elsewhere on your system, they are not being managed by Homebrew.
> Homebrew strongly recommends that you `brew install` the missing dependencies:
>   brew install [dependency]

Thanks!

**Reply by SMillerDev** [reply]
> Could Brew Install, Brew Doctor, and Brew Missing include an option (perhaps "--locate-dependencies") such that:

No, because it will make maintaining Homebrew near impossible, because there is no way to tell if things are compatible, eventually causing the project to be abandoned and unusable for everyone. Furthermore:

>
    IF the option is selected AND IF the Brew List does not contain a required dependency
        THEN Homebrew executes a simple which [dependency] (or type [dependency]) shell command
        IF the shell command returns any valid result
            THEN Homebrew recognizes the presence of the dependency and does not attempt to install it or otherwise flag it as missing
            ELSE the shell will return a null result (or "Could not find [dependency]" when the type command is used) in which case Homebrew can treat the dependency as missing just as it does now.
    The description text for this proposed option could start with the same warning currently found on the --ignore-dependencies option: "An unsupported Homebrew development flag...".

This depends on homebrew knowing exactly what is contained and provided by every dependency. It doesn't.
If we just run `which <formula-name>` we're going to have a lot of issues with non-matching names, as well as missing libraries.

**Reply by TemporaryEfx** [reply]
Ah, this makes perfect sense, and I appreciate your explanation! TIL that if application X depends on application on Y, the Brew formula for X is aware of only the _**formula-name**_ for Y. This keeps Homebrew lean and efficient, but because formula-name often differs from executable-name, it means that (contrary to my assumption) the install script for X does not have the info it needs to perform a `which [executable-name]` query to locate Y. Thanks again for making the effort to inform and educate me… this is a great community!
